File: tools/chain/lib/snapshot-format.mjs
// snapshot-format.mjs — the snapshot's version token, read from the one file both languages
// read (`tools/chain/snapshot-format.json`). Data-Contract.md §5.2.
//
// ── WHY A MODULE FOR ONE STRING ─────────────────────────────────────────────────────────
//
// Because it was ten strings. `blocktracer/chain-snapshot@1` was spelled at four producers
// (`follow-chain.mjs`, `capture-chain.mjs`, `ingest-range.mjs` twice), two consumers
// (`ingest.nim`'s gate, `backfill-blocks.mjs`'s refusal), one test helper and three committed
// fixtures — and the reader's gate was a bare `!=` against its own copy of the literal. A
// version gate spelled twice is two gates, and the interesting failure is not that one goes
// stale but that the tree can then claim a token whose requirements it does not meet.
//
// That is not hypothetical here. `refusalReason` became MANDATORY on every untraced row
// while the token stayed `@1`, and the evidence that this is a new format rather than an
// additive change is that `migrate-refusal-reasons.mjs` had to be written: the committed
// `@1` captures could not pass the gate their own producers now run. `@1` therefore named
// two incompatible shapes. See `snapshot-format.json` for the full statement.

import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const FORMAT_PATH = join(dirname(fileURLToPath(import.meta.url)), '..', 'snapshot-format.json');

// `readFileSync` rather than a JSON import assertion, for the reason `refusal.mjs` gives:
// the assertion syntax is still flagged on some Node versions this repository runs under,
// and a module that fails to LOAD takes the pipeline down for a reason unrelated to what it
// decides.
const spec = JSON.parse(readFileSync(FORMAT_PATH, 'utf8'));

if (spec.format !== 'blocktracer/snapshot-format@1') {
  throw new Error(
    `${FORMAT_PATH} declares format ${JSON.stringify(spec.format)}, and this module only `
    + `knows blocktracer/snapshot-format@1. Refusing to read a version policy it may not `
    + `understand — a half-read version gate is no gate.`);
}

/** The token a producer in this tree WRITES. */
export const SNAPSHOT_FORMAT = spec.current;

/** Every token this tree can read, in order. A token outside this list is refused by name
 *  and never partially read (Data-Contract.md §3, §5.2). */
export const READABLE_SNAPSHOT_FORMATS = Object.freeze([...spec.readable]);

if (!READABLE_SNAPSHOT_FORMATS.includes(SNAPSHOT_FORMAT)) {
  throw new Error(`${FORMAT_PATH}: \`current\` is ${JSON.stringify(SNAPSHOT_FORMAT)} and it `
    + `is not in \`readable\`. A tree that cannot read what it writes is not a version `
    + `policy.`);
}

export const isReadableSnapshotFormat = (t) => READABLE_SNAPSHOT_FORMATS.includes(t);

/** Does this token require `refusalReason` on every untraced row?
 *
 *  THE WHOLE CONTENT OF THE BUMP. `@1` left it optional and `@2` requires it, and the
 *  reader enforces exactly this difference — which is what makes the token a checkable
 *  statement about the tree rather than a label. A producer that writes the member must
 *  write the token that requires it; a tree carrying `@1` is read with the member optional,
 *  fully, with nothing skipped. */
export const requiresRefusalReason = (t) =>
  (spec.mandatoryMembers?.[t] ?? []).includes('transactions[].refusalReason');

/** The three populations a transaction row can be in. Single-sourced here because the
 *  mandatory-member gate above is only checkable against a definition of "untraced", and a
 *  second definition of that is a second answer. `lib/refusal.mjs` reads these. */
export const SNAPSHOT_OUTCOMES = Object.freeze({
  traced: Object.freeze([...spec.outcomes.traced]),
  untraced: Object.freeze([...spec.outcomes.untraced]),
  chainAbsent: Object.freeze([...spec.outcomes.chainAbsent]),
});

// The three lists must be DISJOINT and that is asserted rather than assumed: an outcome in
// two of them would make `accountedFor` double-count and would make "every untraced row
// carries a reason" true and false of the same row.
{
  const seen = new Set();
  for (const [group, members] of Object.entries(SNAPSHOT_OUTCOMES)) {
    for (const o of members) {
      if (seen.has(o)) {
        throw new Error(`${FORMAT_PATH}: outcome ${JSON.stringify(o)} appears in more than `
          + `one population (last seen in ${group}). The three are a partition; an overlap `
          + `makes one of the published counts a double-count.`);
      }
      seen.add(o);
    }
  }
}

/**
 * Refuse a token this tree cannot read, BY NAME.
 *
 * §3: "the site-generator and the conformance validator refuse a version they do not support
 * rather than misreading it." §5.2: the reader "refuses a token it does not know, by name,
 * rather than attempting a partial read". One implementation, so the producers' own
 * consistency checks and the reader say the same thing.
 *
 * @param {string} token the `format` member as found
 * @param {string} where the file or tool, for the message
 */
export function assertReadableSnapshotFormat(token, where) {
  if (isReadableSnapshotFormat(token)) return token;
  throw new Error(
    `${where}: unsupported chain snapshot format ${JSON.stringify(token ?? null)}; this `
    + `build reads ${READABLE_SNAPSHOT_FORMATS.join(' and ')}. Refused by name rather than `
    + `read in part — a snapshot half-read against the wrong schema publishes a chain that `
    + `never existed.`);
}

File: src/blocktracer/chain/snapshot_format.nim
## chain/snapshot_format.nim
##
## The reader's half of the snapshot's version policy — Data-Contract.md §5.2.
##
## ## Why this file reads bytes instead of declaring a string
##
## For the reason `refusal_reasons.nim` beside it gives, and the cost here is the
## same shape. The producer's half is `tools/chain/lib/snapshot-format.mjs`; both
## halves read `tools/chain/snapshot-format.json`, this one with `staticRead` at
## COMPILE time, so a token added there is readable here without anybody
## remembering to copy it and a malformed policy fails the BUILD rather than
## producing a gate that waves everything through.
##
## The gate was previously a bare `!=` against a literal in `ingest.nim`, with
## the same literal spelled at four producers, two consumers, one test helper and
## three committed fixtures. What that arrangement cannot express is precisely
## what happened: `refusalReason` became MANDATORY on every untraced row while
## every one of those ten sites went on saying `@1`, so the token named two
## incompatible shapes and nothing in an artifact could say which it was.
##
## ## What `@2` is, and what the reader does with `@1`
##
## `@2` requires `transactions[].refusalReason` on every untraced row. `@1` left
## it optional, and this reader is the one that treated it as optional — which is
## why the bump is not cosmetic and also why `@1` stays readable: the only member
## `@2` adds is one this side already tolerated the absence of.
##
## §3's rule is that a version the reader does not support is refused BY NAME
## rather than misread, and §5.2's that an unknown token is never partially read.
## `@1` is neither unknown nor partially read — it is enumerated in `readable`
## and every member of it is consumed. What changes is that on `@2` the reader
## ENFORCES the member, which is what makes the token a checkable statement about
## the tree rather than a label on it.
##
## ## The outcome partition is here because the gate needs it
##
## "Every untraced row carries a reason" is only checkable against a definition
## of untraced, and the three populations are not interchangeable: a traced row
## must carry NO reason id, an untraced one must carry one, and a chain-absent
## row (Aztec's private half) must carry a SENTENCE and no id at all. The
## producer side reads the same three lists out of the same file.

import std/[json, strutils]

const snapshotFormatJson = staticRead("../../../tools/chain/snapshot-format.json")

type
  SnapshotFormatPolicy* = object
    current*: string             ## the token a producer in this tree writes
    readable*: seq[string]       ## every token this build can read, in order
    mandatoryRefusalReason*: seq[string]
      ## the tokens on which `transactions[].refusalReason` is required
    traced*: seq[string]
    untraced*: seq[string]
    chainAbsent*: seq[string]

proc parseSnapshotFormat(): SnapshotFormatPolicy =
  let doc = parseJson(snapshotFormatJson)
  if doc{"format"}.getStr != "blocktracer/snapshot-format@1":
    raise newException(ValueError,
      "tools/chain/snapshot-format.json declares format '" &
      doc{"format"}.getStr & "', and this module only knows " &
      "blocktracer/snapshot-format@1. A half-read version gate is no gate.")
  result.current = doc{"current"}.getStr
  if result.current.len == 0:
    raise newException(ValueError,
      "tools/chain/snapshot-format.json names no `current` token.")
  for t in doc{"readable"}.getElems:
    let s = t.getStr
    if s.len == 0:
      raise newException(ValueError, "an empty token in `readable`")
    result.readable.add s
  if result.readable.len == 0:
    raise newException(ValueError,
      "tools/chain/snapshot-format.json lists no readable tokens. An empty list " &
      "would refuse every snapshot ever written, including this tree's own.")
  if result.current notin result.readable:
    raise newException(ValueError,
      "tools/chain/snapshot-format.json: `current` is '" & result.current &
      "' and it is not in `readable`. A tree that cannot read what it writes is " &
      "not a version policy.")
  let mandatory = doc{"mandatoryMembers"}
  if mandatory == nil or mandatory.kind != JObject:
    raise newException(ValueError,
      "tools/chain/snapshot-format.json declares no `mandatoryMembers`. Without " &
      "it the tokens differ by nothing a reader can check, which is the state " &
      "this file exists to end.")
  for tok in result.readable:
    let members = mandatory{tok}
    if members == nil or members.kind != JArray:
      raise newException(ValueError,
        "tools/chain/snapshot-format.json: readable token '" & tok &
        "' has no `mandatoryMembers` entry. A token whose requirements are " &
        "unstated is a token whose gate cannot be written — state the empty " &
        "list deliberately.")
    for m in members.getElems:
      if m.getStr == "transactions[].refusalReason":
        result.mandatoryRefusalReason.add tok
  let outcomes = doc{"outcomes"}
  if outcomes == nil or outcomes.kind != JObject:
    raise newException(ValueError,
      "tools/chain/snapshot-format.json declares no `outcomes` partition.")
  for (group, dest) in [("traced", 0), ("untraced", 1), ("chainAbsent", 2)]:
    let arr = outcomes{group}
    if arr == nil or arr.kind != JArray or arr.len == 0:
      raise newException(ValueError,
        "tools/chain/snapshot-format.json: outcome population '" & group &
        "' is missing or empty. An empty population makes every claim over it " &
        "vacuously true, which is the shape of green this whole campaign is about.")
    for o in arr.getElems:
      case dest
      of 0: result.traced.add o.getStr
      of 1: result.untraced.add o.getStr
      else: result.chainAbsent.add o.getStr
  # THE THREE ARE A PARTITION, and an overlap is a build failure rather than a
  # double-count at run time. An outcome in two populations would make
  # `accountedFor` count it twice and would make "every untraced row carries a
  # reason" both required and forbidden of the same row.
  var seen: seq[string]
  for o in result.traced & result.untraced & result.chainAbsent:
    if o in seen:
      raise newException(ValueError,
        "tools/chain/snapshot-format.json: outcome '" & o & "' appears in more " &
        "than one population. The three are a partition.")
    seen.add o

const
  SnapshotFormat* = parseSnapshotFormat()
    ## The version policy, read at compile time from the file the producers read.

proc currentSnapshotFormat*(): string = SnapshotFormat.current

proc isReadableSnapshotFormat*(token: string): bool =
  token in SnapshotFormat.readable

proc readableSnapshotFormatList*(): string =
  ## The tokens, for a refusal message. A refusal that names only what it did not
  ## recognise makes the reader go looking; one that also names what it accepts
  ## makes the fix visible from the failure.
  SnapshotFormat.readable.join(" or ")

proc snapshotRequiresRefusalReason*(token: string): bool =
  ## Does this token require a `refusalReason` on every untraced row?
  ##
  ## THE WHOLE CONTENT OF THE `@1` → `@2` BUMP. On `@1` the reader accepts a row
  ## without one, which is the behaviour that let the pre-ING-3 captures be read;
  ## on `@2` an untraced row without one is refused, naming the row. A version
  ## whose only difference a reader does not act on is a label.
  token in SnapshotFormat.mandatoryRefusalReason

proc isUntracedSnapshotOutcome*(outcome: string): bool =
  outcome in SnapshotFormat.untraced

proc isChainAbsentSnapshotOutcome*(outcome: string): bool =
  outcome in SnapshotFormat.chainAbsent

proc isTracedSnapshotOutcome*(outcome: string): bool =
  outcome in SnapshotFormat.traced
